pkg/credentialprovider/keyring.go

```diff
@@ -77,17 +81,20 @@ func (dk *BasicDockerKeyring) Add(cfg DockerConfig) {
 		//    foo.bar.com/namespace
 		// Or hostname matches:
 		//    foo.bar.com
+		// It also considers /v2/  and /v1/ equivalent to the hostname
 		// See ResolveAuthConfig in docker/registry/auth.go.
-		if parsed.Host != "" {
-			// NOTE: foo.bar.com comes through as Path.
-			dk.creds[parsed.Host] = append(dk.creds[parsed.Host], creds)
-			dk.index = append(dk.index, parsed.Host)
+		effectivePath := parsed.Path
+		if strings.HasPrefix(effectivePath, "/v2/") || strings.HasPrefix(effectivePath, "/v1/") {
+			effectivePath = effectivePath[3:]
 		}
-		if (len(parsed.Path) > 0) && (parsed.Path != "/") {
-			key := parsed.Host + parsed.Path
-			dk.creds[key] = append(dk.creds[key], creds)
-			dk.index = append(dk.index, key)
+		var key string
+		if (len(effectivePath) > 0) && (effectivePath != "/") {
+			key = parsed.Host + effectivePath
+		} else {
+			key = parsed.Host
 		}
+		dk.creds[key] = append(dk.creds[key], creds)
```

Comment:
Previously, /v1/ and /v2/ were being treated as answering to `host.name`, `host.name/v1/` and `host.name/v2/`.  That means that an auth url like "host.name/v2/some/repo" would be resolved to matching `host.name`, which is very incorrect (we should never present credentials for that).  This change specifically handles those paths and tightens.